OpenRA.Mods.AS/Activities/EnterNydus.cs

```diff
@@ -0,0 +1,110 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2015- OpenRA.Mods.AS Developers (see AUTHORS)
+ * This file is a part of a third-party plugin for OpenRA, which is
+ * free software. It is made available to you under the terms of the
+ * GNU General Public License as published by the Free Software
+ * Foundation. For more information, see COPYING.
+ */
+#endregion
+
+using System.Drawing;
+using OpenRA.Mods.AS.Traits;
+using OpenRA.Mods.Common.Activities;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.AS.Activities
+{
+	class EnterNydus : Enter
+	{
+		public EnterNydus(Actor self, Actor target, EnterBehaviour enterBehaviour)
+			: base(self, target, enterBehaviour)
+		{
+		}
+
+		protected override bool CanReserve(Actor self)
+		{
+			// Primary building is where you come out!
+			return !Target.Actor.IsPrimaryNydusExit();
+		}
+
+		protected override void OnInside(Actor self)
+		{
+			// entered the nydus canal but the entrance is dead immediately. haha;;
+			if (Target.Actor.IsDead)
+				return;
+
+			// Find the primary nydus exit.
+			var pri = self.Owner.PlayerActor.Trait<NydusCounter>().PrimaryActor;
+
+			// Unfortunately, primary exit is killed for some reason and not exists at this time.
+			if (pri == null)
+				return;
```

Comment:
Wait, what? This probably makes the unit lost for the player, that's definitely a no-go.

We need to check if we can kick back the enterer to the world here.